[Autotuner] Feature: add `--cpu_budget` and `--timeout_per_trial`

luarss opened the pull request:
Rationale
- More intuitive for end users who might just want to specify the total number of cpus to be blocked.
- `timeout_per_trial` is different from overall `timeout`. 

TODO
- [x] Tests for `--cpu_budget` -> verify the timeout is hit.
- [x] Tests for timeout per trial and timeout -> does it actually stop the entire training?

luarss committed:
add initial scaffold

Signed-off-by: Jack Luar <jluar@precisioninno.com>

luarss committed:
fix args.job syntax error

Signed-off-by: Jack Luar <jluar@precisioninno.com>

luarss committed:
add smoke test timeout

Signed-off-by: Jack Luar <jluar@precisioninno.com>

luarss committed:
consistent file names and fix smoke_test_sweep

Signed-off-by: Jack Luar <jluar@precisioninno.com>

luarss committed:
move resources_per_trial to sweep_parser

Signed-off-by: Jack Luar <jluar@precisioninno.com>

luarss committed:
fix timeout test and add cpu_budget logic

Signed-off-by: Jack Luar <jluar@precisioninno.com>

luarss committed:
add `--yes` to avoid prompt.

Signed-off-by: Jack Luar <jluar@precisioninno.com>

luarss committed:
add prototype for CPUBudgetSmokeTest

Signed-off-by: Jack Luar <jluar@precisioninno.com>

luarss committed:
fix timeout

Signed-off-by: Jack Luar <jluar@precisioninno.com>

luarss committed:
reintroduce args.jobs, and recalculate timeouts accordingly

Signed-off-by: Jack Luar <jluar@precisioninno.com>

luarss committed:
test_trial_timeout -> test_timeout

Signed-off-by: luarss <jluar@precisioninno.com>

luarss committed:
fix cpu_budget default to be None and reduce timeout fail

Signed-off-by: Jack Luar <jluar@precisioninno.com>

luarss committed:
use cpu_budget default -1 to mean unlimited

Signed-off-by: Jack Luar <jluar@precisioninno.com>

luarss committed:
fix smoke timeout

Signed-off-by: Jack Luar <jluar@precisioninno.com>

luarss committed:
fix test timeout by timeout check in subprocess.run, also fix typo in cpubudget prompt from hrs->seconds

Signed-off-by: Jack Luar <jluar@precisioninno.com>

vvbandeira commented on docs/user/InstructionsForAutoTuner.md:
````diff
@@ -133,6 +133,7 @@ Example:
 ```shell
 python3 distributed.py --design gcd --platform sky130hd \
                        --config ../../../../flow/designs/sky130hd/gcd/autotuner.json \
+                       --yes \
````
We should prompt first users so that they know what is happening.

vvbandeira commented on docs/user/InstructionsForAutoTuner.md:
````diff
@@ -144,6 +145,7 @@ Example:
 ```shell
 python3 distributed.py --design gcd --platform sky130hd \
                        --config distributed-sweep-example.json \
+                       --yes \
````
Same as above.

vvbandeira commented on tools/AutoTuner/src/autotuner/distributed.py:
```diff
@@ -863,23 +912,44 @@ def parse_arguments():
+            template = calculate_expected_numbers(overall_timeout, 1)
+        print(template)
+        if not args.yes:
+            ans = input("Are you sure you want to proceed? (y/n): ")
```
Add a print before this line or include in the input statement that the `-y/--yes` is available if the user does not wish to be asked again.

vvbandeira reviewed (changes requested):
Just an early feedback.

luarss committed:
add tooltip, update readmes

Signed-off-by: Jack Luar <jluar@precisioninno.com>

luarss committed:
fix black

Signed-off-by: Jack Luar <jluar@precisioninno.com>

luarss committed:
update test suite

Signed-off-by: Jack Luar <jluar@precisioninno.com>

luarss committed:
add --yes flag

Signed-off-by: Jack Luar <jluar@precisioninno.com>

luarss committed:
do not exit code

Signed-off-by: Jack Luar <jluar@precisioninno.com>

luarss committed:
fixes to smoke_test_timeout
- subprocess bug: check must be False to capture nonzero retcode.
- METRIC key not inside best_result if timeout prematurely called before trial completes (true for smoke test)

Signed-off-by: Jack Luar <jluar@precisioninno.com>

luarss committed:
fix test_tune_resume

Signed-off-by: Jack Luar <jluar@precisioninno.com>

luarss committed:
shorten timeout

Signed-off-by: Jack Luar <jluar@precisioninno.com>

luarss committed:
fix subprocess invocation

Signed-off-by: Jack Luar <jluar@precisioninno.com>

luarss committed:
more generous timeout for resume check

Signed-off-by: Jack Luar <jluar@precisioninno.com>

luarss committed:
uniquify experiment dir

Signed-off-by: Jack Luar <jluar@precisioninno.com>

luarss committed:
Fix experiment name checks

Signed-off-by: Jack Luar <jluar@precisioninno.com>

luarss committed:
cpubudget test fixes
- make a more reasonable # of samples to avoid jenkins worker overloading.
- fix expected_timeout

Signed-off-by: Jack Luar <jluar@precisioninno.com>

luarss committed:
fix lint

Signed-off-by: Jack Luar <jluar@precisioninno.com>

luarss committed:
fix global args, restore ihp smoke test tune

Signed-off-by: Jack Luar <jluar@precisioninno.com>

luarss committed:
remove cpubudget from sweep test

Signed-off-by: Jack Luar <jluar@precisioninno.com>

luarss committed:
standardise cpubudget test

Signed-off-by: Jack Luar <jluar@precisioninno.com>

luarss committed:
standardise ref_file resume tests

Signed-off-by: Jack Luar <jluar@precisioninno.com>

luarss committed:
standardise and fix smoke_test_timeout

Signed-off-by: Jack Luar <jluar@precisioninno.com>

luarss committed:
fix typo

Signed-off-by: Jack Luar <jluar@precisioninno.com>
